File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Java/17-分账/SplitBill.java
package com.java.demo;

import com.java.utils.WXPayUtility; // 引用微信支付工具库，参考：https://pay.weixin.qq.com/doc/v3/partner/4014985777

import com.google.gson.annotations.SerializedName;
import com.google.gson.annotations.Expose;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.HashMap;
import java.util.Map;

/**
 * 申请分账账单（服务商模式）
 */
public class SplitBill {
  private static String HOST = "https://api.mch.weixin.qq.com";
  private static String METHOD = "GET";
  private static String PATH = "/v3/profitsharing/bills";

  public static void main(String[] args) {
    SplitBill client = new SplitBill(
      "YOUR_MCHID",
      "YOUR_CERT_SERIAL_NO",
      "/path/to/apiclient_key.pem",
      "YOUR_PUB_KEY_ID",
      "/path/to/wxp_pub.pem"
    );

    SplitBillRequest request = new SplitBillRequest();
    request.subMchid = "1900000109";
    request.billDate = "2019-06-11";
    request.tarType = SplitBillTarType.GZIP;
    try {
      SplitBillResponse response = client.run(request);
      System.out.println(response);
    } catch (WXPayUtility.ApiException e) {
      e.printStackTrace();
    }
  }

  public SplitBillResponse run(SplitBillRequest request) {
    String uri = PATH;
    Map<String, Object> args = new HashMap<>();
    args.put("sub_mchid", request.subMchid);
    args.put("bill_date", request.billDate);
    args.put("tar_type", request.tarType);
    String queryString = WXPayUtility.urlEncode(args);
    if (!queryString.isEmpty()) {
      uri = uri + "?" + queryString;
    }

    Request.Builder reqBuilder = new Request.Builder().url(HOST + uri);
    reqBuilder.addHeader("Accept", "application/json");
    reqBuilder.addHeader("Wechatpay-Serial", wechatPayPublicKeyId);
    reqBuilder.addHeader("Authorization", WXPayUtility.buildAuthorization(mchid, certificateSerialNo, privateKey, METHOD, uri, null));
    reqBuilder.method(METHOD, null);
    Request httpRequest = reqBuilder.build();

    OkHttpClient client = new OkHttpClient.Builder().build();
    try (Response httpResponse = client.newCall(httpRequest).execute()) {
      String respBody = WXPayUtility.extractBody(httpResponse);
      if (httpResponse.code() >= 200 && httpResponse.code() < 300) {
        WXPayUtility.validateResponse(this.wechatPayPublicKeyId, this.wechatPayPublicKey,
            httpResponse.headers(), respBody);
        return WXPayUtility.fromJson(respBody, SplitBillResponse.class);
      } else {
        throw new WXPayUtility.ApiException(httpResponse.code(), respBody, httpResponse.headers());
      }
    } catch (IOException e) {
      throw new UncheckedIOException("Sending request to " + uri + " failed.", e);
    }
  }

  private final String mchid;
  private final String certificateSerialNo;
  private final PrivateKey privateKey;
  private final String wechatPayPublicKeyId;
  private final PublicKey wechatPayPublicKey;

  public SplitBill(String mchid, String certificateSerialNo, String privateKeyFilePath, String wechatPayPublicKeyId, String wechatPayPublicKeyFilePath) {
    this.mchid = mchid;
    this.certificateSerialNo = certificateSerialNo;
    this.privateKey = WXPayUtility.loadPrivateKeyFromPath(privateKeyFilePath);
    this.wechatPayPublicKeyId = wechatPayPublicKeyId;
    this.wechatPayPublicKey = WXPayUtility.loadPublicKeyFromPath(wechatPayPublicKeyFilePath);
  }

  public static class SplitBillRequest {
    @SerializedName("sub_mchid")
    @Expose(serialize = false)
    public String subMchid;

    @SerializedName("bill_date")
    @Expose(serialize = false)
    public String billDate;

    @SerializedName("tar_type")
    @Expose(serialize = false)
    public SplitBillTarType tarType;
  }

  public static class SplitBillResponse {
    @SerializedName("hash_type")
    public SplitBillHashType hashType;

    @SerializedName("hash_value")
    public String hashValue;

    @SerializedName("download_url")
    public String downloadUrl;
  }

  public enum SplitBillTarType {
    @SerializedName("GZIP")
    GZIP
  }

  public enum SplitBillHashType {
    @SerializedName("SHA1")
    SHA1
  }

}

File: skills/wechatpay-basic-payment/references/2-服务商/示例代码/Go/17-分账/SplitBill.go
package main

import (
	"demo/wxpay_utility" // 引用微信支付工具库，参考 https://pay.weixin.qq.com/doc/v3/partner/4015119446
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
)

func main() {
	config, err := wxpay_utility.CreateMchConfig(
		"YOUR_MCHID",
		"YOUR_CERT_SERIAL_NO",
		"/path/to/apiclient_key.pem",
		"YOUR_PUB_KEY_ID",
		"/path/to/wxp_pub.pem",
	)
	if err != nil {
		fmt.Println(err)
		return
	}

	request := &SplitBillRequest{
		SubMchid: wxpay_utility.String("1900000109"),
		BillDate: wxpay_utility.String("2019-06-11"),
		TarType:  SPLITBILLTARTYPE_GZIP.Ptr(),
	}

	response, err := SplitBill(config, request)
	if err != nil {
		fmt.Printf("请求失败: %+v\n", err)
		return
	}

	fmt.Printf("请求成功: %+v\n", response)
}

func SplitBill(config *wxpay_utility.MchConfig, request *SplitBillRequest) (response *SplitBillResponse, err error) {
	const (
		host   = "https://api.mch.weixin.qq.com"
		method = "GET"
		path   = "/v3/profitsharing/bills"
	)

	reqUrl, err := url.Parse(fmt.Sprintf("%s%s", host, path))
	if err != nil {
		return nil, err
	}
	query := reqUrl.Query()
	if request.SubMchid != nil {
		query.Add("sub_mchid", *request.SubMchid)
	}
	if request.BillDate != nil {
		query.Add("bill_date", *request.BillDate)
	}
	if request.TarType != nil {
		query.Add("tar_type", fmt.Sprintf("%v", *request.TarType))
	}
	reqUrl.RawQuery = query.Encode()
	httpRequest, err := http.NewRequest(method, reqUrl.String(), nil)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Accept", "application/json")
	httpRequest.Header.Set("Wechatpay-Serial", config.WechatPayPublicKeyId())
	authorization, err := wxpay_utility.BuildAuthorization(config.MchId(), config.CertificateSerialNo(), config.PrivateKey(), method, reqUrl.RequestURI(), nil)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Authorization", authorization)

	client := &http.Client{}
	httpResponse, err := client.Do(httpRequest)
	if err != nil {
		return nil, err
	}
	respBody, err := wxpay_utility.ExtractResponseBody(httpResponse)
	if err != nil {
		return nil, err
	}
	if httpResponse.StatusCode >= 200 && httpResponse.StatusCode < 300 {
		err = wxpay_utility.ValidateResponse(config.WechatPayPublicKeyId(), config.WechatPayPublicKey(), &httpResponse.Header, respBody)
		if err != nil {
			return nil, err
		}
		response := &SplitBillResponse{}
		if err := json.Unmarshal(respBody, response); err != nil {
			return nil, err
		}
		return response, nil
	} else {
		return nil, wxpay_utility.NewApiException(httpResponse.StatusCode, httpResponse.Header, respBody)
	}
}

type SplitBillRequest struct {
	SubMchid *string           `json:"sub_mchid,omitempty"`
	BillDate *string           `json:"bill_date,omitempty"`
	TarType  *SplitBillTarType `json:"tar_type,omitempty"`
}

func (o *SplitBillRequest) MarshalJSON() ([]byte, error) {
	type Alias SplitBillRequest
	a := &struct {
		SubMchid *string           `json:"sub_mchid,omitempty"`
		BillDate *string           `json:"bill_date,omitempty"`
		TarType  *SplitBillTarType `json:"tar_type,omitempty"`
		*Alias
	}{
		SubMchid: nil,
		BillDate: nil,
		TarType:  nil,
		Alias:    (*Alias)(o),
	}
	return json.Marshal(a)
}

type SplitBillResponse struct {
	HashType    *SplitBillHashType `json:"hash_type,omitempty"`
	HashValue   *string            `json:"hash_value,omitempty"`
	DownloadUrl *string            `json:"download_url,omitempty"`
}

type SplitBillTarType string

func (e SplitBillTarType) Ptr() *SplitBillTarType { return &e }

const (
	SPLITBILLTARTYPE_GZIP SplitBillTarType = "GZIP"
)

type SplitBillHashType string

func (e SplitBillHashType) Ptr() *SplitBillHashType { return &e }

const (
	SPLITBILLHASHTYPE_SHA1 SplitBillHashType = "SHA1"
)
